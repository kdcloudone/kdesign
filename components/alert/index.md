---
category: Components
type: 反馈
order: 0
title: Alert
subtitle: 警告提示
---

警告提示是一种用于页面中展示提示信息的组件。

## 使用场景

需要用户关注提示信息时。

## API

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| banner | 是否悬浮在顶部 | boolean | `false` | `true` `false` | 1.0.0 |
| bannerOffset | 悬浮在顶部时的位置设置 | Array<left: number, top: number> | `[0, 0]` | - | 1.0.0 |
| className | 传入反馈浮层组件的样式 | String | - | - | 1.0.0 |
| closable | 是否可以手动关闭反馈浮层 | boolean | `false` | `true` `false` | 1.0.0 |
| closeNode | 自定义的关闭元素 | ReactNode | - | - | 1.0.0 |
| delayOffTime | 反馈浮层显示的时间 | number | `5000` | - | 1.0.0 |
| extra | 关闭按钮前自定义元素 | ReactNode | - | - | 1.0.0 |
| icon | 反馈浮层的自定义信息标识 | ReactNode | - | - | 1.0.0 |
| message | 反馈浮层的显示文本/自定义元素 | ReactNode | - | - | 1.0.0 |
| onClose | 反馈浮层关闭回调 | Function | - | - | 1.0.0 |
| showIcon | 反馈浮层是否显示信息标识图标 | boolean | `false` |  `true` `false` | 1.0.0 |
| type | 反馈浮层的类型 | string | `warning` | `success` `warning` `error` `info` | 1.0.0 |