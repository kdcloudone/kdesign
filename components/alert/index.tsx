import Alert from './alert'
export { AlertType, IAlertProps } from './alert'
export default Alert
