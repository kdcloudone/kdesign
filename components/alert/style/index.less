@import '../../style/themes/index';
@import './mixin.less';
@alert-prefix-cls: ~'@{kd-prefix}-alert';
@icon-side-length: 16px;
@font-size: 12px;
@line-height: 18px;

@success: success;
@warning: warning;
@error: error;
@info: info;

.alert-bg-type-set() {
  @success: @success-alert-bg-color;
  @warning: @warning-alert-bg-color;
  @error: @error-alert-bg-color;
  @info: @info-alert-bg-color;
};

.alert-font-type-set() {
  @success: @success-color;
  @warning: @warning-color;
  @error: @error-color;
  @info: @link-color;
};

.icon-center {
  i {
    align-self: center;
  }
}

.@{alert-prefix-cls} {
  &-container {
    display: none;
    margin: 8px 0;
    min-height: 50px;
    width: 100%;
    // flex-grow: 1;
    align-items: center;
    z-index: @zIndex-apex;
     transition: display 3s;
  }
  &-banner-container {
    display: flex;
    flex-direction: column;
    position: fixed;
    z-index: @zIndex-apex;
    width: 100%;
  }
  each(.alert-bg-type-set(), {
    .alert-bg-type(@key, @value);
  });
  each(.alert-font-type-set(), {
    .alert-font-type(@key, @value);
  });
  &-icon {
    display: flex;
    margin: 0 8px 0 32px;
    .square(@icon-side-length);
    cursor: default;
    .icon-center();
  }
  &-message {
    margin-right: 51px;
    font-size: @font-size;
    line-height: @line-height;
    display: flex;
    flex-grow: 1;
    cursor: default;
  }
  &-without-icon-message {
    margin-left: 56px;
  }
  &-leave {
    transition: min-height @transition-duration @alert-transition-fn, opacity @transition-duration @alert-transition-fn, margin @transition-duration @alert-transition-fn, height @transition-duration @alert-transition-fn, line-height @transition-duration @alert-transition-fn;
    display: flex;
    min-height: 0;
    margin: 0 !important;
    overflow: hidden;
    opacity: 0;
    
    .@{alert-prefix-cls}-icon, .@{alert-prefix-cls}-message, .@{alert-prefix-cls}-close-icon {
      height: 0;
      line-height: 0;
    }
  }
  &-close-icon {
    display: flex;
    .square(@icon-side-length);
    color: @secondary-text-color;
    justify-self: flex-end;
    margin-right: 15px;
    cursor: pointer;
    &:hover {
      color: @theme-color-hover;
    }
    &:active {
      color: @theme-color-click;
    }
    .icon-center();
  }
  &-visible {
    display: flex;
  }
  &-banner {
    position: fixed;
    animation: moveIn @transition-duration @alert-transition-fn both;
  }
}
@keyframes moveIn {
  from {
    transform: translateY(-50px);
  }
  to {
    transform: translateY(0);
  }
}