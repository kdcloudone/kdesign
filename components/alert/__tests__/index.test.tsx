import React from 'react'
import { mount, render, ReactWrapper } from 'enzyme'
import Alert from '../index'
import { AlertTypes, AlertType, allBannerAlertPortals } from '../alert'
import { act } from 'react-dom/test-utils'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'
jest.useFakeTimers()

describe('Alert', () => {
  mountTest(Alert!)
  AlertTypes.forEach((type) => {
    mountTest(() => <Alert type={type} delayOffTime={0} />)
  })
  it('renders correctly', () => {
    expect(render(<Alert />)).toMatchSnapshot()
    AlertTypes.forEach((type) => {
      expect(render(<Alert type={type} delayOffTime={0} />)).toMatchSnapshot()
    })
  })
  it('warns if type is wrong', () => {
    const mockWarn = jest.fn()
    jest.spyOn(console, 'warn').mockImplementation(mockWarn)
    const type = 'who am I' as any as AlertType
    render(<Alert type={type} delayOffTime={0} />)
    expect(mockWarn).toHaveBeenCalledTimes(1)
    expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-alert: cannot found alert type 'who am I'")
  })

  it('should have displayName static property', () => {
    const wrapper = mount(<Alert type={'success'} delayOffTime={0}></Alert>)
    expect((wrapper.type() as any).displayName).toBe('Alert')
  })

  it('should delayOffTime, is zero, can be stay at original placement; is not zero, can be unvisible after delayOffTime ms', async () => {
    let wrapper: ReactWrapper
    await act(async () => {
      wrapper = mount(<Alert type={'success'} message={'Alert Component1'} delayOffTime={0} />)
    })
    let unvisibleWrapper: ReactWrapper
    await act(async () => {
      unvisibleWrapper = mount(<Alert type={'success'} message={'Alert Component1'} delayOffTime={1000} />)
    })
    expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
    jest.runTimersToTime(1000)
    unvisibleWrapper!.update()
    expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
    expect(unvisibleWrapper!.exists('.kd-alert-visible')).toBe(false)
    unvisibleWrapper!.setProps({
      delayOffTime: 0,
    })
    unvisibleWrapper!.update()
    expect(unvisibleWrapper!.exists('.kd-alert-visible')).toBe(true)
    wrapper!.setProps({
      delayOffTime: 1000,
    })
    expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
    jest.runTimersToTime(1000)
    wrapper!.update()
    expect(wrapper!.exists('.kd-alert-visible')).toBe(false)
  })

  it('should showIcon & closable use right', () => {
    const showIconClosableArray = [
      [true, true],
      [true, false],
      [false, true],
      [false, false],
    ]
    showIconClosableArray.forEach(([showIcon, closable]) => {
      expect(
        render(<Alert type={'success'} showIcon={showIcon} closable={closable} delayOffTime={0} />),
      ).toMatchSnapshot()
    })
  })

  it('should class use right', () => {
    const wrapper = mount(
      <Alert type={'success'} message={'Alert Component'} delayOffTime={0} className={'class-test'} />,
    )
    expect(wrapper.exists('.class-test')).toBe(true)
  })

  it('should banner & bannerOffset use right', () => {
    const wrapper = mount(
      <Alert type={'success'} message={'Alert Component'} delayOffTime={0} banner className={'banner-test-a'} />,
    )
    const containerleft0Top0 = document.body.querySelector('.kd-alert-banner-container')
    expect(getComputedStyle(containerleft0Top0!, null).getPropertyValue('left')).toBe('0px')
    expect(getComputedStyle(containerleft0Top0!, null).getPropertyValue('top')).toBe('0px')
    expect(containerleft0Top0?.children.length).toBe(1)
    expect(allBannerAlertPortals.size).toBe(1)
    const otherWrapper = mount(
      <Alert type={'success'} message={'Alert Component1'} delayOffTime={0} banner className={'banner-test-b'} />,
    )
    expect(containerleft0Top0?.children.length).toBe(2)
    expect(allBannerAlertPortals.size).toBe(1)
    expect(containerleft0Top0?.children[0].classList.contains('banner-test-a')).toBe(true)
    expect(containerleft0Top0?.children[1].classList.contains('banner-test-b')).toBe(true)
    otherWrapper.setProps({
      bannerOffset: [10, 10],
    })
    otherWrapper.update()
    expect(document.body.querySelectorAll('.kd-alert-banner-container').length).toBe(2)
    expect(allBannerAlertPortals.size).toBe(2)
    const containerleft10Top10 = document.body.querySelectorAll('.kd-alert-banner-container')[1]
    expect(containerleft10Top10?.children[0].classList.contains('banner-test-b')).toBe(true)
    expect(containerleft0Top0?.children.length).toBe(1)
    expect(containerleft10Top10?.children.length).toBe(1)
    expect(getComputedStyle(containerleft10Top10!, null).getPropertyValue('left')).toBe('10px')
    expect(getComputedStyle(containerleft10Top10!, null).getPropertyValue('top')).toBe('10px')
    wrapper.setProps({
      bannerOffset: [10, 20],
    })
    wrapper.update()
    // containerleft0Top0 to be removed
    jest.runAllTimers()
    const containerleft10Top20 = document.body.querySelectorAll('.kd-alert-banner-container')[1]
    expect(containerleft10Top20?.children[0].classList.contains('banner-test-a')).toBe(true)
    expect(containerleft10Top10?.children.length).toBe(1)
    expect(containerleft10Top20?.children.length).toBe(1)
    expect(containerleft0Top0?.children.length).toBe(0)
    expect(document.body.querySelectorAll('.kd-alert-banner-container').length).toBe(2)
    expect(allBannerAlertPortals.size).toBe(2)
    expect(getComputedStyle(containerleft10Top10!, null).getPropertyValue('left')).toBe('10px')
    expect(getComputedStyle(containerleft10Top10!, null).getPropertyValue('top')).toBe('10px')
    expect(getComputedStyle(containerleft10Top20!, null).getPropertyValue('left')).toBe('10px')
    expect(getComputedStyle(containerleft10Top20!, null).getPropertyValue('top')).toBe('20px')
  })

  it('should icon use right', () => {
    const iconFC = function () {
      return <div className="icon-customed" />
    }
    const wrapper = mount(
      <Alert type={'success'} message={'Alert Component'} delayOffTime={0} banner icon={iconFC()} />,
    )
    expect(wrapper.exists('.icon-customed')).toBe(false)
    wrapper.setProps({
      showIcon: true,
    })
    wrapper.update()
    expect(wrapper.exists('.icon-customed')).toBe(true)
  })

  it('should closeNode use right', () => {
    const closeNodeFC = function () {
      return <div className="closeNode-customed" />
    }
    const wrapper = mount(
      <Alert type={'info'} message={'Alert Component'} delayOffTime={0} banner closeNode={closeNodeFC()} />,
    )
    expect(wrapper.exists('.closeNode-customed')).toBe(false)
    wrapper.setProps({
      closable: true,
    })
    wrapper.update()
    expect(wrapper.exists('.closeNode-customed')).toBe(true)
  })

  it('should extra use right', () => {
    const extraFC = function () {
      return <div className="extra-customed" />
    }
    const wrapper = mount(
      <Alert type={'success'} message={'Alert Component'} delayOffTime={0} banner extra={extraFC()} />,
    )
    expect(wrapper.exists('.extra-customed')).toBe(true)
  })

  it('should config use config provider', () => {
    const alertConfig = {
      compDefaultProps: {
        Alert: {
          type: 'success',
        },
      },
    }
    const wrapper = mount(
      <ConfigProvider value={alertConfig}>
        <Alert delayOffTime={0}></Alert>
      </ConfigProvider>,
    )
    expect(wrapper.find('.kd-alert-container')).toHaveClassName('.alert-success-bg-color.alert-success-font-color')
  })

  it('test click', () => {
    const onClose = jest.fn()
    const wrapper = mount(<Alert type={'info'} showIcon closable onClose={onClose} delayOffTime={0} />)
    act(() => {
      wrapper.find('.kd-alert-close-icon').simulate('click')
    })
    expect(onClose).toHaveBeenCalled()
  })

  it('test mouseover & mouseout', async () => {
    const onClose = jest.fn()
    let wrapper: ReactWrapper
    await act(async () => {
      wrapper = mount(
        <Alert type={'success'} message={'test mouseover & mouseout'} showIcon closable onClose={onClose} />,
      )
    })
    jest.runTimersToTime(4900)
    expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
    await act(async () => {
      wrapper!.find('.kd-alert-container').simulate('mouseover')
    })
    jest.runTimersToTime(5000)
    wrapper!.update()
    expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
    await act(async () => {
      wrapper.find('.kd-alert-container').simulate('mouseout')
    })
    jest.runTimersToTime(5000)
    wrapper!.update()
    expect(wrapper!.exists('.kd-alert-visible')).toBe(false)
    expect(onClose).not.toHaveBeenCalled()
  })
})
