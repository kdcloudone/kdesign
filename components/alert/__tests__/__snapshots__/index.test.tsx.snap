// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders correctly 1`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-warning-bg-color alert-warning-font-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert renders correctly 2`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-success-bg-color alert-success-font-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert renders correctly 3`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-info-bg-color alert-info-font-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert renders correctly 4`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-warning-bg-color alert-warning-font-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert renders correctly 5`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-error-bg-color alert-error-font-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert should showIcon & closable use right 1`] = `
<div
  class="kd-alert kd-alert-closable kd-alert-leave kd-alert-container alert-success-bg-color alert-success-font-color"
>
  <div
    class="kd-alert-icon"
  >
    <i
      class="kdicon kdicon-right-solid"
    />
  </div>
  <div
    class="kd-alert-message"
  />
</div>
`;

exports[`Alert should showIcon & closable use right 2`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-success-bg-color alert-success-font-color"
>
  <div
    class="kd-alert-icon"
  >
    <i
      class="kdicon kdicon-right-solid"
    />
  </div>
  <div
    class="kd-alert-message"
  />
</div>
`;

exports[`Alert should showIcon & closable use right 3`] = `
<div
  class="kd-alert kd-alert-closable kd-alert-leave kd-alert-container alert-success-bg-color alert-success-font-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert should showIcon & closable use right 4`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-success-bg-color alert-success-font-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;
